src/OpenTelemetry/Trace/Configuration/OpenTelemetrySdk.cs

```diff
@@ -66,10 +67,10 @@ public static void EnableOpenTelemetry(Action<OpenTelemetryBuilder> configureOpe
 
                 // Function which takes ActivitySource and returns true/false to indicate if it should be subscribed to
                 // or not
-                ShouldListenTo = (activitySource) => openTelemetryBuilder.ActivitySourceNames.Contains(activitySource.Name.ToUpperInvariant()),
+                ShouldListenTo = (activitySource) => openTelemetryBuilder.ActivitySourceNames?.Contains(activitySource.Name.ToUpperInvariant()) ?? false,
 
                 // The following parameter is not used now.
-                GetRequestedDataUsingParentId = (ref ActivityCreationOptions<string> options) => ActivityDataRequest.AllData,
+                GetRequestedDataUsingParentId = (ref ActivityCreationOptions<string> options) => ActivityDataRequest.None, // Defer to parent whether or not we should record.
```

Comment:
@cijothomas I think we had a bug here. Before we returned `ActivityDataRequest.AllData` which would turn on sampling for a child when its parent was not being sampled. The way I'm reading the source (https://github.com/dotnet/runtime/blob/bd4697644e7024d1e5be94b674f73066bee112a4/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs#L123-L126) we need to return less than the parent in order to preserve its sampling decision?

@tarekgh API is a bit confusing here. Maybe return should be `null` when you want to defer to the parent's current `ActivityDataRequest` mode? Also, only the Id is available from the parent, would we ever need the full parent Activity here?

/cc @MikeGoldsmith @pjanotti

Reply:
Nevermind! Reverted the change. I realized writing a unit test for this that Activity w/ parent still flows through `GetRequestedDataUsingContext`. Once https://github.com/dotnet/runtime/pull/37185 is available I'll add that test. Can't do it right now unless this code is changed to pass `Activity.Current?.Context ?? default` as `parentContext` when starting its Activity, which is what the PR is addressing. All that being said, we'll have Parent.ActivityTraceFlags at time of sampling (gets us Recorded), but don't we also need Parent.IsAllDataRequested? That's not on `ActivityContext`.

My comment that the API is a bit confusing still stands :)

Reply:
@CodeBlanch 

> API is a bit confusing here. Maybe return should be null when you want to defer to the parent's current ActivityDataRequest mode? Also, only the Id is available from the parent, would we ever need the full parent Activity here?

The GetRequestedDataUsingParentId callback is called only when anyone call Activity.StartActivity and passing the parent Id. Originally this callback requested by @noahfalk and will be used in the scenario in which the parent Id is enough to make the decision. I don't think the parent Activity object will be available in such scenarios, otherwise, we would have the API taking the parent Activity instead of the parent Id. The question now is, why you think the API is confusing? and what exactly the suggestion here? the listener has full control here to decide if interested in sampling such Activity with that parent Id or just ignore it and later decide about the child activities. why this is not enough?

Reply:
@tarekgh Sorry this got long. I tried to make it "not a rambling mess" but the topic is big.

> why you think the API is confusing?

Hard to say exactly! Sorry I know that isn't helpful :) We have the concept of "sampling" which I think people are naturally familiar with (somewhat?) but nothing in ActivitySource is really named along the lines of that. We have `GetRequestedDataUsingParentId` & `GetRequestedDataUsingContext`. Mental leap there to relate "GetRequestedData" to sampling and what we're trying to accomplish. I think just renaming these `GetSamplingModeUsing...` (GetSamplingLevel?) would help a lot. Maybe rename `ActivityDataRequest` -> `ActivitySamplingMode`. Just my $.02!

Here's an example of how hard it is to use this correctly:

Imagine a request comes into an application and an Activity is started. Sampler logic is run. It decides to sample at `ActivityDataRequest.PropagationData` level, for whatever reason. Now we go to call some dependency. A new Activity is created. Sampler looks at this new one. All it gets from the Parent (Activity.Current.Context) is ActivityTraceFlags, which will be `ActivityTraceFlags.None` in the case of `ActivityDataRequest.PropagationData` when Parent was sampled. What is a sampler to do here? It might think... the parent isn't recorded, nor should I be. Or it might think, I can decide to sample if I want to. But really it needs to know if ActivityContext != default than the parent was `ActivityDataRequest.PropagationData` and it should probably do the same for this child?

To fix that up I think we need to keep `ActivityDataRequest` on the Activity (or ActivityContext) and pass that in to `GetRequestedDataUsing...` for children. That will be maximum clarity.

Check out our current implementation:
https://github.com/open-telemetry/opentelemetry-dotnet/blob/dcedf02ec5d94f19826c6c42b897b9bbb6d8df69/src/OpenTelemetry/Trace/Configuration/OpenTelemetrySdk.cs#L84-L104

It doesn't have the rule for `PropagationData` so for my exercise above, we won't propagate the child. Bug?

It should be:
```csharp
                    if (shouldSample.IsSampled)
                    {
                        return ActivityDataRequest.AllDataAndRecorded;
                    }
                    else if (options.Parent != default)
                    {
                        return ActivityDataRequest.PropagationData;
                    }
                    else
                    {
                        return ActivityDataRequest.None;
                    }
```
No?

Reply:
There seem to be two separate suggestions here:
1. Better name for RequestedData - like SamplingMode
2. Make the ActivityDataRequest of parent available to `GetRequestedDataUsingContext` callback, so child can know the exact parent decision - instead of just knowing ActivityContext

Both are relevant, but not tied to this PR alone - lets discuss in separate issue, and proceed with merging this PR. 

(I can create issues or @CodeBlanch can - let me know)